File: src/ros2_bridge/mappers/sensor_msgs/camera_info.rs
//! Typed mapper for `sensor_msgs/msg/CameraInfo`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn camera_info_to_bus(msg: ros_env::sensor_msgs::msg::CameraInfo) -> crate::sensor_msgs::msg::v1::CameraInfo {
    crate::sensor_msgs::msg::v1::CameraInfo {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        height: msg.height.into(),
        width: msg.width.into(),
        distortion_model: crate::ros2_bridge::mappers::convert::from_ros_string(msg.distortion_model),
        d: crate::ros2_bridge::mappers::convert::f64_seq(msg.d),
        k: crate::ros2_bridge::mappers::convert::f64_seq(msg.k),
        r: crate::ros2_bridge::mappers::convert::f64_seq(msg.r),
        p: crate::ros2_bridge::mappers::convert::f64_seq(msg.p),
        binning_x: msg.binning_x.into(),
        binning_y: msg.binning_y.into(),
        roi: Some(crate::ros2_bridge::mappers::sensor_msgs::region_of_interest::region_of_interest_to_bus(msg.roi)),
    }
}

pub(crate) fn camera_info_to_ros(bus: crate::sensor_msgs::msg::v1::CameraInfo) -> ros_env::sensor_msgs::msg::CameraInfo {
    ros_env::sensor_msgs::msg::CameraInfo {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        height: bus.height as _,
        width: bus.width as _,
        distortion_model: crate::ros2_bridge::mappers::convert::to_ros_string(bus.distortion_model),
        d: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.d),
        k: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.k),
        r: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.r),
        p: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.p),
        binning_x: bus.binning_x as _,
        binning_y: bus.binning_y as _,
        roi: crate::ros2_bridge::mappers::sensor_msgs::region_of_interest::region_of_interest_to_ros(bus.roi.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsCameraInfoMapper;

impl TypedTopicMapper for SensorMsgsCameraInfoMapper {
    type Ros = ros_env::sensor_msgs::msg::CameraInfo;
    type Bus = crate::sensor_msgs::msg::v1::CameraInfo;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(camera_info_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(camera_info_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/camera_info.py
"""Generated mapper for `sensor_msgs/msg/CameraInfo`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.sensor_msgs.region_of_interest import region_of_interest_to_bus, region_of_interest_to_ros

def camera_info_to_bus(msg):
    from robot_bus.sensor_msgs.msg.v1 import CameraInfo as BusMsg

    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.height = msg.height
    bus.width = msg.width
    bus.distortion_model = str(msg.distortion_model)
    bus.d.extend(list(msg.d))
    bus.k.extend(list(msg.k))
    bus.r.extend(list(msg.r))
    bus.p.extend(list(msg.p))
    bus.binning_x = msg.binning_x
    bus.binning_y = msg.binning_y
    bus.roi.CopyFrom(region_of_interest_to_bus(msg.roi))
    return bus


def camera_info_to_ros(bus):
    from sensor_msgs.msg import CameraInfo as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.height = bus.height
    out.width = bus.width
    out.distortion_model = str(bus.distortion_model)
    out.d = list(bus.d)
    out.k = list(bus.k)
    out.r = list(bus.r)
    out.p = list(bus.p)
    out.binning_x = bus.binning_x
    out.binning_y = bus.binning_y
    out.roi = region_of_interest_to_ros(bus.roi)
    return out


class SensorMsgsCameraInfoMapper:
    def type_name(self) -> str:
        return "sensor_msgs/msg/CameraInfo"

    def ros_msg_type(self):
        from sensor_msgs.msg import CameraInfo as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return camera_info_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.sensor_msgs.msg.v1 import CameraInfo as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return camera_info_to_ros(bus)
